### ISSUE TITLE
Expression adapting

### ISSUE THREAD

**Issue by mrahhal** [mrahhal/MR.EntityFrameworkCore.KeysetPagination#37]
This issue is about an enhancement of how we parse and work with the configured keyset column building expressions.

Right now we have support for two particular cases:
- Simple access: `x => x.Prop`
- Nested access: `x => x.Nested.Prop`

This issue is about completely overhauling how we deal with these expressions. Instead of supporting particular syntax, we should rewrite internal expressions to adapt it for use when building the keyset and other usages (i.e obtaining a value from a reference).

Error reporting should also be better and clearer for when there is incompatibility or unsupported syntax.

The goal is to also allow the following, among other patterns, when `Created` is nullable:
```cs
b.Ascending(x => x.Created ?? DateTime.MinValue)
```

This is instead of having to create a computed column to produce the same result. Note however that an indexed computed column will probably always be the preferred method for performance.

The following should also be possible as long as the resulting expression is non nullable.
```cs
b.Ascending(x => x.Created ?? x.Updated)
```

*Note*: This might harm performance as the query is forcing a scan instead of a seek. Is this something we should support in the first place? Since keyset pagination is all about performance.

Might also be related to https://github.com/mrahhal/MR.EntityFrameworkCore.KeysetPagination/issues/33.

**Comment by mrahhal** [comment]
It's important to note that any solution that doesn't use a computed column might sacrifice performance. This is because you could index a computed column, whereas something like a `COALESCE("x"."Created", '0001-01-01 00:00:00')` (which is the translated `x.Created ?? DateTime.MinValue`) will not be properly indexed and so you end up with a table scan instead of a seek. 
While I was able to, in theory, get `x.Created ?? DateTime.MinValue` to work, this is especially problematic in this library because this library only exists to be a faster method of pagination. So now I'm doubting the benefit of allowing such dynamic expressions.

**Comment by d00ML0rDz** [comment]
Good to hear you've had some success with this. I was having a little look at our project where we use MR.EntityFrameworkCore.KeysetPagination in, and we have situations where we need to use a ternary expression where I don't think using a computed column is the right thing to do performance wise.

An example is where we use nested objects. Here we have two types of profiles, user profiles and business profiles. Posts can have both a UserProfile and a BusinessProfile associated with them. If we want to sort all our profiles by the amount of posts they have, it would look something like this:

```csharp
queryable.KeysetPaginate(p => p.Descending(profile => profile.Type == EProfileType.Business ? profile.BusinessProfile.Posts.Count : profile.UserProfile.Posts.Count));
```

However, we don't really want a computed column getting a profile's post count every time we load a profile as it would only be necessary for this one situation. Maybe you know of a better way to handle this, but the solution you've proposed in #37 looks like it would solve this?

**Comment by mrahhal** [comment]
Thanks for providing a specific example. Yes, in theory the expression parsing improvement does allow this to work (although this is a bit more complicated than the usual cases so I'll have to check), but as mentioned this won't have good performance characteristics, which in my opinion defeats the purpose of using keyset pagination in the first place. The only advantage of keyset pagination is performance.

> However, we don't really want a computed column getting a profile's post count every time we load a profile as it would only be necessary for this one situation

If the computed column is indexed (or has `PERSISTED`) then it will update when the underlying data changes, and won't be calculated on each query (although this does mean the db will have more work on inserts). Even if it doesn't, depending on how you load these (if you use DTOs for example), you can avoid querying this computed column in the cases you don't need it in.

In any case, this is a bit of an advanced case (count). Didn't think of this particular usage before. I would definitely still prefer having an indexed computed column for this for the best performance. Can you tell me what you're doing right now in this particular case you showed?

**Comment by d00ML0rDz** [comment]
From what I understand, keyset pagination also has the benefit of not skipping or showing duplicate rows if records higher in the order get added or removed like is the case with offset pagination, at least when sorting by a static value. That makes it pretty important for us as we have a lot of infinite scroll newsfeeds in our applications which heavily benefit from that and your library makes that much easier to handle.

In the example I gave above for the profile's post counts, that is a small feature I wanted to add in our admin application. Currently, we have a feed of all the profiles that exist on our platform which originally was only sorted by their Created DateTime property, but I wanted to add a way to sort them by the amount of posts among a few other options. This feature is currently built on MR.EntityFrameworkCore.KeysetPagination already, with us having a few other sorting options such as the amount of logins and followers etc, hence why the expression parsing solution would be a quick and easy solution for us. It was more of a nice to have feature rather than something important, so I didn't want to look at going down the route of setting up computed columns.

This admin app is currently only used by two people, so performance isn't too important in this case. I appreciate in our consumer facing application we should go down the route of computed columns.

**Comment by mrahhal** [comment]
> From what I understand, keyset pagination also has the benefit of not skipping or showing duplicate rows if records higher in the order get added or removed like is the case with offset pagination, at least when sorting by a static value.

That's true in the normal case. But since you have a dynamically changing column in your keyset (count), pagination stability (no skipping or duplication) can't be guaranteed. Stability is there if your keyset's data doesn't change (maybe that's what you meant by static).

But now I realize that maybe there's a use case for a non-indexed keyset because of this stability characteristic. My only problem then is finding a way to make it clear to the user when their keyset might not be what they want in terms of performance. As it currently stands, if I allow any expression, it'll be easy for a user to think they'll get better perf because of keyset pagination but they're actually getting less because of their non-indexed keyset.

> This feature is currently built on MR.EntityFrameworkCore.KeysetPagination already, with us having a few other sorting options such as the amount of logins and followers etc

But so you don't have this particular `profile.Type == EProfileType.Business ? profile.BusinessProfile.Posts.Count : profile.UserProfile.Posts.Count` equivalent? Just wondering if you had a way to workaround this particular case.

**Comment by d00ML0rDz** [comment]
> (maybe that's what you meant by static)

Sorry, that is what I meant by static 😊

I'm no expert at this sort of thing, but would it be possible to create a warning in the IDE to say performance would be impacted if using a complex expression rather than a computed column? That might be one way to go about it + some guidance in your repo/nuget package README?

![image](https://user-images.githubusercontent.com/66832778/235425355-700ca65f-5fbb-4b69-bad0-11c456bb7337.png)

Currently I don't have a workaround for `profile.Type == EProfileType.Business ? profile.BusinessProfile.Posts.Count : profile.UserProfile.Posts.Count`, so I've not added it to our admin application yet.

**Comment by mrahhal** [comment]
Using the analyzer to show a warning is my first thought towards a possible solution, but it might be a bit complicated to avoid false positives. Need to try a few things out.

I see. Would definitely like to unblock these cases if I have a good method in place to warn the user (which then they can suppress). Thanks for the example.

**Comment by d00ML0rDz** [comment]
Sounds good, just let me know if I can give a hand with testing any changes. Thank you!

**Comment by mrahhal** [comment]
This will be coming soon with the next version release. All the examples listed should now work.

**Comment by mrahhal** [comment]
Released in v1.3.0.

**Comment by d00ML0rDz** [comment]
Just had the chance to test these changes and they're working perfectly, thank you very much 😄 Really appreciated!

**Comment by mrahhal** [comment]
Great! Thanks for confirming.
